### ISSUE TITLE
how to understand the loss function corner case for 0 and 255

### ISSUE THREAD

**Issue by weixsong** [openai/pixel-cnn#23]
Hi,

In the loss function, the code give a special value to value 0 and 255:
```python
    # log probability for edge case of 0 (before scaling)
    log_cdf_plus = plus_in - tf.nn.softplus(plus_in)
    # log probability for edge case of 255 (before scaling)
    log_one_minus_cdf_min = -tf.nn.softplus(min_in)
```

I could not understand these code, how could me understand it?

**Comment by npuichigo** [comment]
I think you should know the relationship between sigmoid and softplus.
![image](https://user-images.githubusercontent.com/11533479/34025235-b3b1768e-e189-11e7-99cd-91bcf503826b.png)
![image](https://user-images.githubusercontent.com/11533479/34025241-b9ce55a0-e189-11e7-97db-7412d5be1618.png)

**Comment by npuichigo** [comment]
I think you should know the relationship between sigmoid and softplus.
![image](https://user-images.githubusercontent.com/11533479/34025304-18722870-e18a-11e7-8085-60d697e2b9f0.png)

**Comment by weixsong** [comment]
@npuichigo , thanks very much. I got it.

**Comment by npuichigo** [comment]
By the way, could you understand these code for robustness?
```python
mid_in = inv_stdv * centered_x
    # log probability in the center of the bin, to be used in extreme cases
    # (not actually used in our code)
    log_pdf_mid = mid_in - log_scales - 2. * tf.nn.softplus(mid_in)
```
I'm not quite sure what these code mean.

**Comment by chediBechikh** [comment]
Hi all,
the explanation is not clear here and what relation what exactly used for this line

 # log probability for edge case of 0 (before scaling)
    log_cdf_plus = plus_in - tf.nn.softplus(plus_in)
